src/broadcast/broadcast.rs

```diff
@@ -284,21 +365,88 @@ impl<N: NodeIdT> Broadcast<N> {
             // Enqueue a broadcast of a Ready message.
             step.extend(self.send_ready(hash)?);
         }
+        // Upon receiving 2f + 1 matching Ready(h) messages, send full
+        // `Echo` message to every node who hasn't sent us a `CanDecode`
+        if self.count_readys(hash) == 2 * self.netinfo.num_faulty() + 1 {
+            // `send_echo` function should take care of which nodes to send the message
+            step.extend(self.send_echo(self.value_message.clone())?);
+        }
+
         Ok(step.join(self.compute_output(hash)?))
     }
 
     /// Sends an `Echo` message and handles it. Does nothing if we are only an observer.
     fn send_echo(&mut self, p: Proof<Vec<u8>>) -> Result<Step<N>> {
-        self.echo_sent = true;
-        if !self.netinfo.is_validator() {
+       if !self.netinfo.is_validator() {
             return Ok(Step::default());
         }
         let echo_msg = Message::Echo(p.clone());
-        let step: Step<_> = Target::All.message(echo_msg).into();
+        let mut step = Step::default();
+        // TODO: iterator not correctly describing right of our_id. Use cycle method on
+        // iterator to correctly specify.
+        let right = self.netinfo.all_ids().skip(self.netinfo.num_correct()+self.pessimissm_factor);
+
+        // Send `Echo` to all left nodes.
+        if !self.echo_sent_partial {
+            self.echo_sent_partial = true;
+            for id in right {
+                let msg = Target::Node(id.clone()).message(echo_msg.clone());
+                step.messages.push(msg);
+            }
+        }
+        // Send `Echo` to remaining right nodes who haven't sent us `CanDecode`.
+        else {
+            self.echo_sent_full = true;
+            for id in right {
+                if !self.can_decodes.contains_key(id) {
+                    let msg = Target::Node(id.clone()).message(echo_msg.clone());
+                    step.messages.push(msg);
+                }
+            }
+        }
```

Comment:
Instead of making this distinction, it might be cleaner to just have two methods: `send_echo_to_left` and `send_echo_to_rest` or something (maybe with better names :sweat_smile:). The former would be called in `handle_value` and the latter in `handle_ready`. Neither of them needs an argument: Currently `self.value_message` is cloned and passed in at both call sites.

Reply:
Yup I agree its pretty messy right now :sweat_smile: 
Will create separate functions as you suggested :)